### PR TITLE
MM-56876: Redis: first introduction

### PR DESCRIPTION
**Pull request by agnivade**
Adding initial version of Redis into the codebase. Some features:
- Selected github.com/redis/rueidis as the Redis client rather than the popular one because of several built-in features: including automatic pipelining of concurrent requests, and client-side caching. Both of which have played a massive role in performance.
- Introduced a new  `MGET` cache API to take advantage of making a single request to get values for multiple keys. This is to reduce network roundtrip.
- We take client-side cache of 16MiB per connection, and expiry time of 5 minutes. Can be tuned later.
- Better logging when reading from cache fails. Previously we never logged the error, and directly fetched from DB. This is why we never got any errors from Redis initially.

FUTURE THINGS TO DO:
- Add docker image and CI tests
- Work on changing KEYS to SCAN
- Some more direct updates
- System console UI
- Migrate status, session, ProfilesInChannel cache as well.

https://mattermost.atlassian.net/browse/MM-56876

```release-note
Added experimental support for Redis. This is not intended for production use yet.
```

### CODE REVIEW

File: server/channels/store/localcachelayer/role_layer.go
Comment (agnivade) [inline server/channels/store/localcachelayer/role_layer.go:85]: This was a MASSIVE bug which I uncovered. Even after we got all roles from cache, we would make an empty call to the DB, get no results anyways, and then return the cached results. This would basically make the cache useless for this call. Making this change should have a big impact.

Actually, I will send this improvement separately. This is not related to Redis at all. And this is also why we keep seeing high number of store calls to the Rolelayer. Because the cache layer doesn't work at all!

Comment (streamer45) [reply server/channels/store/localcachelayer/role_layer.go:85]: :man_facepalming: great find!

Comment (agnivade) [reply server/channels/store/localcachelayer/role_layer.go:85]: Actually, it's not a huge problem .. because of 

```go
func (s *SqlRoleStore) GetByNames(names []string) ([]*model.Role, error) {
	if len(names) == 0 {
		return []*model.Role{}, nil
	}
...
```

So the problem is that while we are making the calls to the store layer, which is shooting up the Grafana charts, we aren't _really_ making a DB call. So while it's good to do that change, there's no urgency to send it separately. This can go as part of this PR itself.

File: server/channels/store/localcachelayer/role_layer.go
Comment (streamer45) [inline server/channels/store/localcachelayer/role_layer.go:70]: A couple of thoughts here, but they apply to the other occurrences of this pattern.

1. Not sure I fully understood the need for pointer of pointer.
2. Minor thing, but we could use `make()` and preallocate since we know the length in advance.

File: server/channels/store/localcachelayer/role_layer.go
Comment (streamer45) [reply server/channels/store/localcachelayer/role_layer.go:85]: Good we had that failsafe :sweat_smile:

File: server/channels/store/localcachelayer/role_layer.go
Comment (streamer45) [inline server/channels/store/localcachelayer/role_layer.go:79]: Trying to understand whether this can ever be `nil`. If so, should we handle it as a miss or else?

File: server/platform/services/cache/provider.go
Comment (streamer45) [inline server/platform/services/cache/provider.go:93]: Do we have an idea of how many connections will be used? Is there a hard cap or something?

File: server/platform/services/cache/redis.go
Comment (streamer45) [inline server/platform/services/cache/redis.go:66]: `time.Since(now).Seconds()`?

File: server/platform/services/cache/redis.go
Comment (streamer45) [inline server/platform/services/cache/redis.go:107]: Should this duration become a constant or something a little more explicit?

File: server/platform/services/cache/redis.go
Comment (streamer45) [inline server/platform/services/cache/redis.go:129]: Would `reflect` be helpful here? I know we are generally trying to avoid it but the use case here seems to fit.

**Review by streamer45 — COMMENTED** [review]
Great work! I like how the whole implementation turned out to be rather clean. I left some comments and thoughts after a quick first pass.

**Review by isacikgoz — COMMENTED** [review]
This was long waited, great job @agnivade 🚀 I have a question on a loop iteration but other than that nothing serious. BTW, are TODOs will be addressed with this PR? Or you'll be creating tickets for those?

File: server/channels/store/localcachelayer/channel_layer.go
Comment (agarciamontoro) [inline server/channels/store/localcachelayer/channel_layer.go:248]: Nit: Can this (and the other `toPass` variables in this and other files) be of a specific type? In this case, `[]**model.Channel`?

File: server/channels/store/localcachelayer/channel_layer.go
Comment (agarciamontoro) [inline server/channels/store/localcachelayer/channel_layer.go:440]: What does this comment (and the other ones like this) mean?

File: server/platform/services/cache/provider.go
Comment (agarciamontoro) [inline server/platform/services/cache/provider.go:33]: Can we document what the returned string means?

File: server/platform/services/cache/redis.go
Comment (agarciamontoro) [inline server/platform/services/cache/redis.go:141]: I don't really understand why we need this only for these two types. For my own understanding: could you expand a bit on this?

**Review by agarciamontoro — COMMENTED** [review]
This is massive, thank you! :tada: Reading this I've found that I don't understand some things about how we use our caches. The main one being: why do we use pointers to pointers? Can we remove one level of indirection there? This may be obvious, but I don't see it now :thinking:

File: server/channels/store/localcachelayer/layer.go
Comment (agnivade) [reply server/channels/store/localcachelayer/layer.go:486]: We could, but that would make the `GetMulti` method slightly different than `Get`, and moreover, it would push the hit/miss logic to the lower layer. This keeps the API simple and consistent, at the cost of an additional loop. I am considering to keep this as-is. LMK what you think.

File: server/channels/store/localcachelayer/channel_layer.go
Comment (agnivade) [reply server/channels/store/localcachelayer/channel_layer.go:248]: Haha, the common gotcha! No it can't :smile: 

Reason: the `doMultiReadCache` expects `[]any` because it needs to accept slice of any type. And `[]**model.Channel` can't be passed to a method that accepts `[]any` :)

The only way around is to declare `any` instead of `[]any` but then again you have to cast `any` to `[]any` just to be able to pass a typed slice. Not worth it.

File: server/channels/store/localcachelayer/channel_layer.go
Comment (agnivade) [reply server/channels/store/localcachelayer/channel_layer.go:440]: It means that since Redis is a shared cache, now we can actually update the cache directly by incrementing the count, rather than invalidating it across all nodes in the cluster and let each node re-populate the new value.

File: server/platform/services/cache/redis.go
Comment (agnivade) [reply server/platform/services/cache/redis.go:141]: All of this started when I added `msgp` which would auto-generate fast marshal/unmarshal code which didn't use reflect. Now usually, you wouldn't need this and the code from L117-120 would just do the job. But the issue is that the interface is only implemented for `*model.User`, not `**model.User`, which brings us to the comment I mentioned above:

> // 2. Our cache interface passes pointers to empty pointers, and not pointers
	// to values. This is mainly how all our model structs are passed around.
	// It might be technically possible to use values _just_ for hot structs
	// like these and then return a pointer while returning from the cache function,
	// but it will make the codebase inconsistent, and has some edge-cases to take care of.

So that's why we do this. Can we fix this? Sure. But then we need to change a _lot_ of other things across the entire codebase as well.

**Comment by agnivade** [comment]
@isacikgoz - Yes, I will create tickets for all the pending items.

@agarciamontoro - Reason for pointer-to-pointer is explained here: https://github.com/mattermost/mattermost/pull/27752/files#diff-768a67785c3675d55a1512483f9669e6cbe0dd903ddc4a12d903328be56765e3R125-R129

**Comment by agnivade** [comment]
Tickets created:
1. https://mattermost.atlassian.net/browse/MM-59931
2. https://mattermost.atlassian.net/browse/MM-59932
3. https://mattermost.atlassian.net/browse/MM-59933
4. https://mattermost.atlassian.net/browse/MM-59934
5. https://mattermost.atlassian.net/browse/MM-59935

File: server/channels/store/localcachelayer/layer.go
Comment (isacikgoz) [reply server/channels/store/localcachelayer/layer.go:486]: I see your point and I'd be in favor of consistency, but isn't this a hot path that whenever we hit to cache, we have to do an additional loop multiplied by size of the cache. Am I over estimating it's impact?

Comment (agnivade) [reply server/channels/store/localcachelayer/layer.go:486]: It's all in-memory operations. There's no network roundtrip. I believe there's little to gain here.

File: server/channels/store/localcachelayer/channel_layer.go
Comment (agarciamontoro) [reply server/channels/store/localcachelayer/channel_layer.go:248]: Ooooooooh, I see. That's not good :joy:

File: server/platform/services/cache/redis.go
Comment (agarciamontoro) [reply server/platform/services/cache/redis.go:143]: > But the issue is that the interface is only implemented for `*model.User`, not `**model.User`

Ah! That's the thing I was missing. That makes sense. Again, just for the sake of satisfying my curiosity: can we generate another interface for `**model.User`?

File: server/channels/store/localcachelayer/channel_layer.go
Comment (agarciamontoro) [reply server/channels/store/localcachelayer/channel_layer.go:440]: So is this a TODO? Because we're invalidating the cache right below the comment, right? Or am I missing something else?

Comment (agnivade) [reply server/channels/store/localcachelayer/channel_layer.go:440]: Yep, it's a TODO. Ticket already filed here: https://github.com/mattermost/mattermost/pull/27752#issuecomment-2257490365

File: server/platform/services/cache/redis.go
Comment (agnivade) [reply server/platform/services/cache/redis.go:143]: That's an interesting proposition. So the thing is that these methods are auto-generated, I don't want to change that. But I will take a look to see if I can wrap the `*model.User` method with `**model.User`.

**Comment by agnivade** [comment]
Gentle ping for re-review @isacikgoz @agarciamontoro @streamer45

**Comment by streamer45** [comment]
First one that approves wins a prize.

**Comment by agnivade** [comment]
Is this again some trick prize like the AI bot? :stuck_out_tongue_closed_eyes:

File: server/channels/store/localcachelayer/layer.go
Comment (isacikgoz) [reply server/channels/store/localcachelayer/layer.go:486]: Gotcha, if you see the gains negligible it's fine then.

**Comment by agnivade** [comment]
Is everyone waiting for the other one to approve first? The suspense is killing me :laughing: 

(Code freeze is next week btw)

**Review by isacikgoz — APPROVED** [review]
I want the prize!

**Comment by agnivade** [comment]
@streamer45 @agarciamontoro - Now that Ibrahim has won the prize, I wait your approval :)

File: server/platform/services/cache/redis.go
Comment (streamer45) [inline server/platform/services/cache/redis.go:187]: I wonder if there could be a more idiomatic way to do this. Intermixing errors with cache misses has already caused some problems and here we are taking it to another level.

File: server/platform/services/cache/redis.go
Comment (streamer45) [inline server/platform/services/cache/redis.go:46]: I see we are using `context.Background()`  pretty much everywhere in this package. Given these are network requests underneath I wonder whether we should establish some timeout value or is it already captured by the client itself?

File: server/platform/services/cache/redis.go
Comment (streamer45) [inline server/platform/services/cache/redis.go:77]: Random thought: have we considered using a pool of buffers to back this rather than allocating every time?

File: server/platform/services/cache/redis.go
Comment (agnivade) [reply server/platform/services/cache/redis.go:46]: Yep, it is captured by the client itself. https://pkg.go.dev/github.com/redis/rueidis#ClientOption - `ConnWriteTimeout`. It is the read/write timeout for each connection.

It was also advised against setting context based timeouts in the earlier client: https://redis.uptrace.dev/guide/go-redis-debugging.html#timeouts

File: server/platform/services/cache/redis.go
Comment (agnivade) [reply server/platform/services/cache/redis.go:77]: We already use a pool in the `Get` path by using a pool of objects for some caches. Since the `Get` path is much more hit, rather than the `Set` path.

We could use a pool of buffers, but the challenge here is to identify exactly when to return the buffer back to pool.

**Review by streamer45 — APPROVED** [review]

**Comment by agnivade** [comment]
@saturninoabril - Hey Saturn, wondering if there's someone available to just do a quick sanity check on HA spinwick? Dylan is OOO and I'd like to merge this before code freeze.

This doesn't add anything new. Redis is added in a disabled mode. But just ensuring that there is no inadvertent bug here. Just a light sanity check on a HA spinwick should suffice.

**Comment by saturninoabril** [comment]
Tested and no issue found on HA setup (tested at `https://mattermost-pr-27752.test.mattermost.cloud/`, since HA spinwick seemed to have issue particularly running without license and so HA was disabled). Previous E2E test: no issue (https://github.com/mattermost/mattermost/pull/27752#issuecomment-2265453721). 

Other observation: there were multiple logs thrown with the following (could be related here or cache in general, note that it didn't happen on master/release-9.11):
```json
{
  "caller": "api4/config.go:440",
  "level": "warn",
  "method": "PUT",
  "msg": "Unrecognized config permissions tag value.",
  "path": "/api/v4/config/patch",
  "tag_value": "sysconsole_read_cache",
  "timestamp": "2024-08-05 10:28:07.525 Z",
  <other_fields_redacted>
}
```

**Comment by agnivade** [comment]
> Other observation: there were multiple logs thrown with the following (could be related here or cache in general, note that it didn't happen on master/release-9.11):

Great catch @saturninoabril ! It was because the struct tag was unnecessarily there without a front-facing system console setting. I have removed it now. PTAL.

**Review by saturninoabril — APPROVED** [review]
Thanks! Observation verified as resolve.

**Review by agarciamontoro — APPROVED** [review]
:rocket:

**Comment by cwarnermm** [comment]
@agnivade - Is there product or dev docs impact for this PR?

**Comment by agnivade** [comment]
No docs needed for this. It's not meant for external consumption yet.

**Pull request merged**
